File: src/main/java/org/devcon/ticket/CapabilityIssuer.java
package org.devcon.ticket;

import com.alphawallet.attestation.core.AttestationCrypto;
import com.alphawallet.attestation.core.SignatureUtility;
import com.auth0.jwt.JWT;
import com.auth0.jwt.JWTCreator.Builder;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.util.Base64;
import java.util.Date;
import java.util.Set;
import org.bouncycastle.crypto.AsymmetricCipherKeyPair;

/**
 * Issues a long-term JWT to approve-list a specific domain to give it access to the DevCon ticket API.
 * Specifically the website who holds the token will be able to access a set of specific tasks (methods).
 *
 * This is specifically going to allow a third party site to open an iframe to ticket.devcon
 * in order to access the ticket secret stored in the local cache to construct a useDevconTicket request.
 */
public class CapabilityIssuer extends CapabilityCommon {
  private final KeyPair signingKeys;
  private final URL verifierDomain;

  public CapabilityIssuer(AsymmetricCipherKeyPair signingKeys, String verifierDomain) throws MalformedURLException {
    this.signingKeys = SignatureUtility.convertBouncyCastleKeysToJavaKey(signingKeys);
    this.verifierDomain = new URL(verifierDomain);
  }

  public String makeToken(String domain, Set<String> tasks, int expirationTimeInDays) throws MalformedURLException {
    URL urlDomain = new URL(domain);
    String flattenedTasks = flattenSet(tasks);
    long currentTime = System.currentTimeMillis();
    long expirationInMs = currentTime + (long) expirationTimeInDays * 24l * 60l * 60l * 1000l;
    return buildSignedToken(urlDomain, flattenedTasks, expirationInMs, currentTime);
  }

  String buildSignedToken(URL domain, String flattenedTasks, long expirationTimeInMs, long creationTimeInMs) {
    Builder builder = JWT.create();
    // Only withAudience, withSubject, withIssuer, withExpiresAt, withNotBefore and withClaim(tasksClaimName) are required
    builder.withClaim(TasksClaimName, flattenedTasks);
    // Both the issuer and verifier is the same
    builder.withAudience(verifierDomain.toString());
    builder.withIssuer(verifierDomain.toString());
    builder.withSubject(domain.toString());
    builder.withNotBefore(new Date(creationTimeInMs));
    builder.withExpiresAt(new Date(expirationTimeInMs));
    // withIssuedAt and withJWTId are OPTIONAL
    builder.withIssuedAt(new Date(creationTimeInMs));
    builder.withJWTId(getJWTID(domain.toString(), flattenedTasks, expirationTimeInMs, creationTimeInMs));
    return builder.sign(getAlgorithm(signingKeys.getPublic(), signingKeys.getPrivate()));
  }

  String flattenSet(Set<String> tasks) {
    if (tasks.size() == 0) {
      throw new IllegalArgumentException("At least one task must be assigned");
    }
    StringBuilder flattenedList = new StringBuilder();
    for (String task : tasks) {
      String normalizedTask = task.toLowerCase().trim();
      if (normalizedTask.contains(",")) {
        throw new IllegalArgumentException("A task contains a ',' which is not permitted");
      }
      flattenedList.append(normalizedTask).append(',');
    }
    String unprunedResult = flattenedList.toString();
    // Remove trailing ','
    return unprunedResult.substring(0, unprunedResult.length()-1);
  }

  private String getJWTID(String domain, String flattenedTasks, long expirationTime, long currentTime) {
    ByteBuffer toHash = ByteBuffer.allocate(domain.length() + TasksClaimName.length() +
        flattenedTasks.length() + 2 * (Long.SIZE / 8));
    toHash.put(domain.getBytes(StandardCharsets.UTF_8));
    toHash.put(TasksClaimName.getBytes(StandardCharsets.UTF_8));
    toHash.put(flattenedTasks.getBytes(StandardCharsets.UTF_8));
    toHash.putLong(expirationTime);
    toHash.putLong(currentTime);

    byte[] digest = AttestationCrypto.hashWithKeccak(toHash.array());
    return Base64.getEncoder().encodeToString(digest);
  }
}

File: src/main/javascript/crypto/src/libs/CapabilityIssuer.ts
import {KeyPair} from "./KeyPair";
import {getInt64Bytes, stringToArray, uint8arrayToBase64, uint8merge} from "./utils";
import {AttestationCrypto} from "./AttestationCrypto";
const jwt = require('jsonwebtoken');

export class CapabilityIssuer {
    private signingKeys: KeyPair;
    static TasksClaimName: string = "org.devcon.ticket.capability";
    // private verifierDomain: string;

    constructor (private privateKeyOrSecret: string, private verifierDomain: string) {}

    makeToken(domain: string, tasks: string[], expirationTimeInDays: number) {
        let flattenedTasks = this.flattenSet(tasks);
        let currentTime: number = Date.now();
        let expirationInMs = currentTime + expirationTimeInDays * 24 * 60 * 60 * 1000;
        return this.buildSignedToken(domain, flattenedTasks, expirationInMs, currentTime);
    }

    buildSignedToken(domain: string, flattenedTasks: string, expirationTimeInMs: number, creationTimeInMs: number) {
        // Only withAudience, withSubject, withIssuer, withExpiresAt, withNotBefore and withClaim(tasksClaimName) are required
        let jwtData: any = {
            aud: this.verifierDomain, //withAudience
            iss: this.verifierDomain, //withIssuer
            sub: domain, //withSubject
            nbf: new Date(creationTimeInMs).valueOf(), //withNotBefore
            exp: new Date(expirationTimeInMs).valueOf(), //withExpiresAt
            // withIssuedAt and withJWTId are OPTIONAL
            iat: new Date(creationTimeInMs).valueOf(), //withIssuedAt
            jti: this.getJWTID(domain, flattenedTasks, expirationTimeInMs, creationTimeInMs)

        };
        let jwtOptions = {
            //algorithm (default: HS256)
            //expiresIn: expressed in seconds or a string describing a time span zeit/ms.
            // Eg: 60, "2 days", "10h", "7d". A numeric value is interpreted as a seconds count. If you use a string be sure you provide the time units (days, hours, etc), otherwise milliseconds unit is used by default ("120" is equal to "120ms").

            // notBefore: expressed in seconds or a string describing a time span zeit/ms.
            //     Eg: 60, "2 days", "10h", "7d". A numeric value is interpreted as a seconds count. If you use a string be sure you provide the time units (days, hours, etc), otherwise milliseconds unit is used by default ("120" is equal to "120ms").
            //
            // audience
            // issuer
            // jwtid
            // subject
            // noTimestamp
            // header
            // keyid
            // mutatePayload
        };

        jwtData[CapabilityIssuer.TasksClaimName] = flattenedTasks;
        // TODO set algorithm
        // jwt.sign({ foo: 'bar' }, privateKey, { algorithm: 'RS256' }, function(err, token) {
        //     console.log(token);
        // });
        return jwt.sign(jwtData, this.privateKeyOrSecret, jwtOptions);
    }

    flattenSet(tasks: string[]): string {
        if (!tasks.length || tasks.length == 0) {
            throw new Error("At least one task must be assigned");
        }
        let flattenedList: string = '';
        tasks.forEach(task => {
            let normalizedTask: string = task.toLowerCase().trim();
            if (normalizedTask.includes(",")) {
                throw new Error("A task contains a ',' which is not permitted");
            }
            flattenedList += normalizedTask + ',';
        })

        // Remove trailing ','
        return flattenedList.slice(0,-1);
    }

    getJWTID(domain: string, flattenedTasks: string, expirationTime: number, currentTime: number): string {
        let toHash: Uint8Array = uint8merge([
            new Uint8Array(stringToArray(domain)),
            new Uint8Array(stringToArray(CapabilityIssuer.TasksClaimName)),
            new Uint8Array(stringToArray(flattenedTasks)),
            new Uint8Array(stringToArray(flattenedTasks)),
            getInt64Bytes(expirationTime),
            getInt64Bytes(currentTime)
            ]
        );

        let digest = AttestationCrypto.hashWithKeccak(toHash);
        return uint8arrayToBase64(digest);
    }
}
